File: packages/healops_opentelemetry_python/healops_opentelemetry/exporter.py
import json
import time
import requests
import os
from typing import Sequence, Optional, Dict
from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult
from opentelemetry.trace import StatusCode, SpanKind
from .utils import extract_file_path_from_stack

class HealOpsSpanExporter(SpanExporter):
    def __init__(self, api_key: str, service_name: str, endpoint: str = "https://engine.healops.ai/otel/errors"):
        self.api_key = api_key
        self.service_name = service_name
        self.endpoint = endpoint

    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
        if not spans:
            return SpanExportResult.SUCCESS

        # Transform spans
        transformed_spans = [self._transform_span(span) for span in spans]

        payload = {
            "apiKey": self.api_key,
            "serviceName": self.service_name,
            "spans": transformed_spans
        }

        try:
            self._send(payload)
            return SpanExportResult.SUCCESS
        except Exception as e:
            if os.getenv("HEALOPS_DEBUG"):
                 print(f"Failed to export spans: {e}")
            return SpanExportResult.FAILURE

    def shutdown(self) -> None:
        pass

    def _transform_span(self, span: ReadableSpan) -> dict:
        attributes = dict(span.attributes) if span.attributes else {}

        # Check for exception events
        stack_trace = None
        for event in span.events:
            if event.name == "exception":
                # OTel Python convention for exception
                if "exception.stacktrace" in event.attributes:
                    stack_trace = event.attributes["exception.stacktrace"]
                    attributes["exception.stacktrace"] = stack_trace
                if "exception.type" in event.attributes:
                    attributes["exception.type"] = event.attributes["exception.type"]
                if "exception.message" in event.attributes:
                    attributes["exception.message"] = event.attributes["exception.message"]

        # If no stack trace in events, check attributes
        if not stack_trace and span.status.status_code == StatusCode.ERROR:
            stack_trace = attributes.get("error.stack") or attributes.get("stack")
            if stack_trace:
                attributes["exception.stacktrace"] = stack_trace

        # Extract code location if available
        if stack_trace and "code.file.path" not in attributes:
            extracted_path = extract_file_path_from_stack(stack_trace)
            if extracted_path:
                attributes["code.file.path"] = extracted_path
                # We could parse line number too if we want to be more precise
                # but util currently returns just path.

        return {
            "traceId": f"{span.context.trace_id:032x}",
            "spanId": f"{span.context.span_id:016x}",
            "parentSpanId": f"{span.parent.span_id:016x}" if span.parent else None,
            "name": span.name,
            "kind": span.kind.name if hasattr(span.kind, 'name') else str(span.kind),
            "timestamp": int(time.time() * 1000), # Export time
            "startTime": self._ns_to_ms(span.start_time),
            "endTime": self._ns_to_ms(span.end_time),
            "attributes": attributes,
            "events": [
                {
                    "name": event.name,
                    "time": self._ns_to_ms(event.timestamp),
                    "attributes": dict(event.attributes) if event.attributes else {}
                }
                for event in span.events
            ],
            "status": {
                "code": span.status.status_code.value,
                "message": span.status.description
            },
            "resource": dict(span.resource.attributes)
        }

    def _ns_to_ms(self, ns: int) -> int:
        return int(ns / 1e6) if ns else 0

    def _send(self, payload: dict, attempt: int = 1) -> None:
        max_retries = 3
        timeout = 3 # 3 seconds

        try:
            headers = {
                "Content-Type": "application/json",
                "User-Agent": "HealOps-OTel-SDK/1.0"
            }
            response = requests.post(self.endpoint, json=payload, headers=headers, timeout=timeout)
            response.raise_for_status()
        except requests.RequestException as e:
            if attempt < max_retries:
                delay = (2 ** attempt) * 0.1 # Exponential backoff
                time.sleep(delay)
                self._send(payload, attempt + 1)
            else:
                raise e

File: packages/healops_opentelemetry_go/exporter.go
package healops

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"

	"go.opentelemetry.io/otel/sdk/trace"
)

// HealOpsExporter implements trace.SpanExporter
type HealOpsExporter struct {
	apiKey      string
	serviceName string
	endpoint    string
	client      *http.Client
}

// NewHealOpsExporter creates a new HealOpsExporter
func NewHealOpsExporter(apiKey string, serviceName string, endpoint string) *HealOpsExporter {
	return &HealOpsExporter{
		apiKey:      apiKey,
		serviceName: serviceName,
		endpoint:    endpoint,
		client:      &http.Client{Timeout: 10 * time.Second},
	}
}

// ExportSpans exports a batch of spans
func (e *HealOpsExporter) ExportSpans(ctx context.Context, spans []trace.ReadOnlySpan) error {
	if len(spans) == 0 {
		return nil
	}

	payload := map[string]interface{}{
		"apiKey":      e.apiKey,
		"serviceName": e.serviceName,
		"spans":       e.transformSpans(spans),
	}

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	req, err := http.NewRequestWithContext(ctx, "POST", e.endpoint, bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "HealOps-OTel-Go-SDK/1.0")

	resp, err := e.client.Do(req)
	if err != nil {
		if os.Getenv("HEALOPS_DEBUG") != "" {
			fmt.Printf("Failed to export spans: %v\n", err)
		}
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return fmt.Errorf("failed to export spans, status: %s", resp.Status)
	}

	return nil
}

// Shutdown shuts down the exporter
func (e *HealOpsExporter) Shutdown(ctx context.Context) error {
	return nil
}

func (e *HealOpsExporter) transformSpans(spans []trace.ReadOnlySpan) []map[string]interface{} {
	transformed := make([]map[string]interface{}, 0, len(spans))

	for _, span := range spans {
		attributes := make(map[string]interface{})
		for _, kv := range span.Attributes() {
			attributes[string(kv.Key)] = kv.Value.AsInterface()
		}

		events := make([]map[string]interface{}, 0, len(span.Events()))
		for _, event := range span.Events() {
			evtAttrs := make(map[string]interface{})
			for _, kv := range event.Attributes {
				evtAttrs[string(kv.Key)] = kv.Value.AsInterface()
			}
			events = append(events, map[string]interface{}{
				"name":       event.Name,
				"time":       event.Time.UnixNano() / 1e6, // ms
				"attributes": evtAttrs,
			})

			// Check for exception stack trace to extract code info
			if event.Name == "exception" {
				if stackTrace, ok := evtAttrs["exception.stacktrace"].(string); ok {
					e.extractCodeInfo(stackTrace, attributes)
				}
			}
		}

		transformed = append(transformed, map[string]interface{}{
			"traceId":      span.SpanContext().TraceID().String(),
			"spanId":       span.SpanContext().SpanID().String(),
			"parentSpanId": span.Parent().SpanID().String(),
			"name":         span.Name(),
			"kind":         span.SpanKind().String(),
			"timestamp":    time.Now().UnixNano() / 1e6,
			"startTime":    span.StartTime().UnixNano() / 1e6,
			"endTime":      span.EndTime().UnixNano() / 1e6,
			"attributes":   attributes,
			"events":       events,
			"status": map[string]interface{}{
				"code":    span.Status().Code,
				"message": span.Status().Description,
			},
		})
	}

	return transformed
}

// Regex to find file paths in Go stack traces
// Format: /path/to/file.go:123 +0x...
var fileLineRegex = regexp.MustCompile(`\s+([^\s]+:\d+)`)

func (e *HealOpsExporter) extractCodeInfo(stackTrace string, attributes map[string]interface{}) {
	// If code info is already present, don't overwrite
	if _, ok := attributes["code.file.path"]; ok {
		return
	}

	lines := strings.Split(stackTrace, "\n")
	for _, line := range lines {
		// Skip standard library or runtime lines if possible (simple heuristic)
		if strings.Contains(line, "/usr/local/go/src/") || strings.Contains(line, "runtime/") {
			continue
		}

		matches := fileLineRegex.FindStringSubmatch(line)
		if len(matches) > 1 {
			fullPath := matches[1] // /path/to/file.go:123
			parts := strings.Split(fullPath, ":")
			if len(parts) >= 2 {
				filePath := parts[0]
				lineNumStr := parts[1]

				// Ensure it looks like a Go source file
				if !strings.HasSuffix(filePath, ".go") {
					continue
				}

				attributes["code.file.path"] = filePath
				if lineNum, err := strconv.Atoi(lineNumStr); err == nil {
					attributes["code.line.number"] = lineNum
				}

				// Try to extract function name from previous line
				// Go stack traces usually have function name on the line before the file path
				// This is hard to do reliably without iterating carefully, but we can try simple approach
				// or just leave it for now.
				return
			}
		}
	}
}
